tests/processing/test_processor.py

```diff
@@ -622,3 +622,87 @@ def test_healthcheck(tmpdir: py.path.local) -> None:
 
     processor._run_once()
     assert tmpdir.join("health.txt").mtime() == health_mtime
+
+
+def test_processor_pause_with_invalid_message() -> None:
+
+    topic = Topic("topic")
+
+    consumer = mock.Mock()
+    strategy = mock.Mock()
+    factory = mock.Mock()
+    factory.create_with_partitions.return_value = strategy
+
+    processor: StreamProcessor[int] = StreamProcessor(
+        consumer, topic, factory, IMMEDIATE
+    )
+
+    # Subscribe to topic
+    subscribe_args, subscribe_kwargs = consumer.subscribe.call_args
+    assert subscribe_args[0] == [topic]
+
+    # Partition assignment
+    partition = Partition(topic, 0)
+    consumer.tell.return_value = {}
+    assignment_callback = subscribe_kwargs["on_assign"]
+    offsets = {partition: 0}
+    assignment_callback(offsets)
+
+    # Message that we will get from polling
+    message = Message(BrokerValue(0, partition, 0, datetime.now()))
+
+    # Message will be rejected
+    consumer.poll.return_value = message.value
+    strategy.submit.side_effect = MessageRejected()
+    with assert_changes(lambda: int(consumer.pause.call_count), 0, 1):
+        processor._run_once()
+        assert strategy.submit.call_args_list[-1] == mock.call(message)
+        assert processor._StreamProcessor__message == message.value  # type: ignore
+
+        with mock.patch("time.time", return_value=time.time() + 5):
+            processor._run_once()  # Should pause now
+
+    # Consumer is in paused state
+    assert processor._StreamProcessor__is_paused is True  # type: ignore
+    # The same rejected message should be carried over
+    assert processor._StreamProcessor__message is not None  # type: ignore
+
+    # All partitions are paused
+    consumer.paused.return_value = set(p for p in offsets)
+    # Simulate a continuous backpressure state where messages are being rejected
+    strategy.submit.side_effect = MessageRejected()
```

Comment:
I don't think you need to assert these internals of StreamProcessor. 
These are implementation details that make a test more fragile when they leak out.
You can still test the property of the system (break if the consumer is paused and the next message is invalid) without asserting the internals.
Adding the assertions on internals means the test breaks upon refactoring, while having a test passing after a backwards compatible refactoring is an important asset to validate the change.

Reply:
is this on the right line? i think this is ok to induce paused state of the consumer. I agree that `assert is_paused` is not really necessary